File: scratch/concurrency_and_parallelism/parallelism/splitting_computational_task/scala/PythagoreanParallel2.scala
// with interchanged sides (3,4,5) and (4,3,5)
// def pythagoreans(start:Int, end: Int) = for {
//   x <- 1 to end
//   y <- 1 to end
//   z <- start to end
//   if (x*x + y*y == z*z)
// } yield (x,y,z)

// without interchanged sides (3,4,5) and (4,3,5), only (3,4,5)
def pythagoreans(start:Int, end: Int) = for {
  x <- 1 to end
  y <- x to end
  z <- start to end
  if (x*x + y*y == z*z)
} yield (x,y,z)


def pythagoreansParallel2(n: Int) = {
  // Have at least as many threads as the number of cores
  val numberOfCores = Runtime.getRuntime().availableProcessors()
  // println(s"number of cores = ${numberOfCores}, input = ${n}")
  // Having more threads than the number of cores does not help.  
  // As this is an evenly distributed problem, we will create 
  // number of chunks == number of cores.
  val possibleChunks = if (n <= numberOfCores) 1 else numberOfCores
  val evenlyDistributedChunkSize = Math.ceil(n.toFloat / possibleChunks).toInt
  val numberOfChunks = Math.ceil(n.toFloat / evenlyDistributedChunkSize).toInt
  // println(s"number of chunks = ${numberOfChunks}, evenlyDistributedChunkSize = ${evenlyDistributedChunkSize}")
  // Having more chunks is better than having few, because more chunks help keep the cores busy
  // Also, after a certain number of large chunks, having more chunks have little benefit
  (2 to numberOfChunks).scanLeft((1, evenlyDistributedChunkSize)) { (acc, elem) =>
    val (start, end) = acc
    val newStart = end + 1
    val newEnd = newStart + evenlyDistributedChunkSize
    if (elem == numberOfChunks || newEnd >= n) 
      (newStart, n) 
    else 
      (newStart, newEnd)
  }
  .filter { case (start, end) => start <= n }
  .par
  .map { case (start, end) => pythagoreans(start, end) }
  .foldLeft(List[(Int, Int, Int)]()) { case (acc, list) => acc ++ list }
}

def time[T, R](f: Function[T, R]): Function[T, R] = {
  return t => {
    val startTime = System.currentTimeMillis  
    val result = f(t)
    val diff = System.currentTimeMillis - startTime
    println(s"Time Taken = $diff(ms).")
    result
  }
} 

//--------------------< SET I >---------------------------
// Parallelization using Parallel Lists (Results averaged over 5 runs)
// (without interchanged sides (3,4,5) and (4,3,5), only (3,4,5))
// ===================================================================
// println(time(pythagoreansParallel2)(10))   
// println(time(pythagoreansParallel2)(100))  
// println(time(pythagoreansParallel2)(250))  
// println(time(pythagoreansParallel2)(500))  
// println(time(pythagoreansParallel2)(750))  
// println(time(pythagoreansParallel2)(1000)) 
// println(time(pythagoreansParallel2)(1250)) 
// println(time(pythagoreansParallel2)(1500)) 

// Parallelization (Results averaged over 5 runs)
// (without interchanged sides (3,4,5) and (4,3,5), only (3,4,5))
//==================================================================================+
//                                               |  Using Futures | Using Par List  |
//==================================================================================+
// println(time(pythagoreansParallel(10)))   //  |    37.2ms      |     68.6ms      |
// println(time(pythagoreansParallel(100)))  //  |    112ms       |     143.8ms     |
// println(time(pythagoreansParallel(250)))  //  |    274.4ms     |     366.6ms     |
// println(time(pythagoreansParallel(500)))  //  |    635.6ms     |     759ms       |
// println(time(pythagoreansParallel(750)))  //  |    1234.2ms    |     759ms       |
// println(time(pythagoreansParallel(1000))) //  |    1766ms      |     1924.8ms    |
// println(time(pythagoreansParallel(1250))) //  |    2121.4ms    |     2649.4ms    |
// println(time(pythagoreansParallel(1500))) //  |    3655.6ms    |     3815.2ms    |
//==================================================================================+

// Sequential Pythagoreans (Results averaged over 5 runs)
// (without interchanged sides (3,4,5) and (4,3,5), only (3,4,5))
// ======================================================
// println(time(pythagoreans)(10))     // 10.4ms
// println(time(pythagoreans)(100))    // 44.8ms
// println(time(pythagoreans)(250))    // 183ms
// println(time(pythagoreans)(500))    // 469ms
// println(time(pythagoreans)(750))    // 1332.8ms
// println(time(pythagoreans)(1000))   // 2649.6ms
// println(time(pythagoreans)(1250))   // 4917.8ms
// println(time(pythagoreans)(1500))   // 8180.2ms


// Parallelization (Results averaged over 5 runs)
// (with interchanged sides (3,4,5) and (4,3,5))
//==================================================================================+
//                                               |  Using Futures | Using Par List  |
//==================================================================================+
// println(time(pythagoreansParallel(10)))   //  |    31.8ms      |   65.6ms        |
// println(time(pythagoreansParallel(100)))  //  |    108.8ms     |   119ms         |
// println(time(pythagoreansParallel(250)))  //  |    323.2ms     |   246.4ms       |
// println(time(pythagoreansParallel(539)))  //  |    1345ms      |   1471ms        |
// println(time(pythagoreansParallel(750)))  //  |    1095.8ms    |   1307.4ms      |
// println(time(pythagoreansParallel(1000))) //  |    1890.6ms    |   2128.6ms      |
// println(time(pythagoreansParallel(1250))) //  |    1422.5ms    |   1711.6ms      |
// println(time(pythagoreansParallel(1500))) //  |    2194ms      |   2389.2ms      |
//==================================================================================+

// Sequential Pythagoreans (Results averaged over 5 runs)
// (with interchanged sides (3,4,5) and (4,3,5))
// ======================================================
// println(time(pythagoreans)(10))     // 11.6ms
// println(time(pythagoreans)(100))    // 88ms
// println(time(pythagoreans)(250))    // 275.5ms
// println(time(pythagoreans)(500))    // 1753.2ms
// println(time(pythagoreans)(750))    // 5127ms
// println(time(pythagoreans)(1000))   // 13506.6ms
// println(time(pythagoreans)(1250))   // 18704ms
// println(time(pythagoreans)(1500))   // 39720.4ms

println("DONE")

File: scratch/concurrency_and_parallelism/parallelism/splitting_computational_task/java/PythagoreanParallel2.java
import java.util.*;
import java.util.function.*;
import java.util.stream.*;

public class PythagoreanParallel2 {
  
  static <R> Function<Integer, R> time(Function<Integer, R> f) {
    return t -> {
      long startTime = System.currentTimeMillis();
      R result = f.apply(t);
      long diff = System.currentTimeMillis() - startTime;
      System.out.println(String.format("Time Taken = %d(ms).", diff));
      return result;
    };
  }
  
  // with interchanged sides (3,4,5) and (4,3,5)
  static List<List<?>> pythagoreans(int start, int end) {
    List<List<?>> result = new ArrayList<>();
    for (int x = 1; x <= end; x++) {
      for (int y = 1; y <= end; y++) {
        for (int z = start; z <= end; z++) {
          if (x*x + y*y == z*z)
            result.add(Arrays.asList(x, y, z));
        }
      }
    }
    return result;
  }
  
  // without interchanged sides (3,4,5) and (4,3,5), only (3,4,5)
  // static List<List<?>> pythagoreans(int start, int end) {
  //   List<List<?>> result = new ArrayList<>();
  //   for (int x = 1; x <= end; x++) {
  //     for (int y = x; y <= end; y++) {
  //       for (int z = start; z <= end; z++) {
  //         if (x*x + y*y == z*z)
  //           result.add(Arrays.asList(x, y, z));
  //       }
  //     }
  //   }
  //   return result;
  // }
   
  static Stream<List<?>> makeChunks(int n) {
    // Have at least as many threads as the number of cores
    int numberOfCores = Runtime.getRuntime().availableProcessors();
    // System.out.println(String.format("number of cores = %d, input = %d", numberOfCores, n));
    // Having more threads than the number of cores does not help.
    // As this is an evenly distributed problem, we will create
    // number of chunks == number of cores.
    int possibleChunks = (n <= numberOfCores) ? 1 : numberOfCores;
    int evenlyDistributedChunkSize = (int) Math.ceil((float) n / possibleChunks);
    int numberOfChunks = (int) Math.ceil( (float) n / evenlyDistributedChunkSize);
    // System.out.println(String.format("number of chunks = %d, evenlyDistributedChunkSize = %d", numberOfChunks, evenlyDistributedChunkSize));
    // Having more chunks is better than having few, because more chunks help keep the cores busy
    // Also, after a certain number of large chunks, having more chunks have little benefit
    List<List<?>> chunks = new ArrayList<>();
    List<Integer> seed = List.of(1, evenlyDistributedChunkSize);
    for (int x = 2; x <= numberOfChunks; x++) {
      chunks.add(seed);
      int start = seed.get(0);
      int end = seed.get(1);
      int newStart = end + 1;
      int newEnd = newStart + evenlyDistributedChunkSize;
      if (x == numberOfChunks || newEnd >= n)
         seed = List.of(newStart, n);
      else
         seed = List.of(newStart, newEnd);
    }
    return chunks.stream().filter(chunk -> {
      int start = (int) chunk.get(0);
      int end = (int) chunk.get(1);
      return start <= n;
    });
  }
  
  static List<List<?>> pythagoreans(int n) {
    return makeChunks(n)
      .parallel()
      .flatMap(chunkBounds -> {
        int start = (int) chunkBounds.get(0);
        int end = (int) chunkBounds.get(1);
        return pythagoreans(start, end).stream();
      })
      .collect(Collectors.toList()); 
  }

  public static void main(String[] args) {
    System.out.println(time(PythagoreanParallel2::pythagoreans).apply(1500));
    System.out.println("DONE");
  }
}

//--------------------< SET I >---------------------------
// Parallelization (Results averaged over 5 runs)
// (without interchanged sides (3,4,5) and (4,3,5), only (3,4,5))
//=======================================================================+
//                                               | Using Parallel Stream |
//=======================================================================+
// println(time(pythagoreansParallel(10)))   //  |        21ms           |
// println(time(pythagoreansParallel(100)))  //  |        22.4ms         |
// println(time(pythagoreansParallel(250)))  //  |        47.2ms         |
// println(time(pythagoreansParallel(500)))  //  |        68.6ms         |
// println(time(pythagoreansParallel(750)))  //  |        112.6ms        |
// println(time(pythagoreansParallel(1000))) //  |        196.2ms        |
// println(time(pythagoreansParallel(1250))) //  |        322.8ms        |
// println(time(pythagoreansParallel(1500))) //  |        500.4ms        |
//=======================================================================+

// Sequential Pythagoreans (Results averaged over 5 runs)
// (without interchanged sides (3,4,5) and (4,3,5), only (3,4,5))
// ======================================================
// println(time(pythagoreans)(10))     // 0ms
// println(time(pythagoreans)(100))    // 3ms
// println(time(pythagoreans)(250))    // 11.8ms
// println(time(pythagoreans)(500))    // 53.4ms
// println(time(pythagoreans)(750))    // 149.8ms
// println(time(pythagoreans)(1000))   // 367.2ms
// println(time(pythagoreans)(1250))   // 684.8ms
// println(time(pythagoreans)(1500))   // 1232ms


// Parallelization (Results averaged over 5 runs)
// (with interchanged sides (3,4,5) and (4,3,5))
//=======================================================================+
//                                               | Using Parallel Stream |
//=======================================================================+
// println(time(pythagoreansParallel(10)))   //  |        20.6ms         |
// println(time(pythagoreansParallel(100)))  //  |        28.8ms         |
// println(time(pythagoreansParallel(250)))  //  |        56.4ms         |
// println(time(pythagoreansParallel(500)))  //  |        93ms           |
// println(time(pythagoreansParallel(750)))  //  |        182ms          |
// println(time(pythagoreansParallel(1000))) //  |        334ms          |
// println(time(pythagoreansParallel(1250))) //  |        609.2ms        |
// println(time(pythagoreansParallel(1500))) //  |        995.2ms        |
//=======================================================================+

// Sequential Pythagoreans (Results averaged over 5 runs)
// (with interchanged sides (3,4,5) and (4,3,5))
// ======================================================
// time(pythagoreans)(10)     // 0ms
// time(pythagoreans)(100)    // 6.6ms
// time(pythagoreans)(250)    // 36.4ms
// time(pythagoreans)(500)    // 251.8ms
// time(pythagoreans)(750)    // 848ms
// time(pythagoreans)(1000)   // 2020.6ms
// time(pythagoreans)(1250)   // 4093ms
// time(pythagoreans)(1500)   // 7163.2ms

